tools/jest_teamcity_reporter/jest_teamcity_reporter.js

```diff
@@ -0,0 +1,298 @@
+// Copyright (c) 2024 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+'use strict'
+
+const assert = require('assert')
+const path = require('path')
+const util = require('util')
+
+/**
+ * @typedef {import('@jest/types').Config.GlobalConfig} GlobalConfig
+ * @typedef {import('@jest/reporters').Reporter} Reporter
+ * @typedef {import('@jest/reporters').Test} Test
+ * @typedef {import('@jest/reporters').TestResult} TestResult
+ */
+
+/** @implements {Reporter} */
+class TeamcityReporter {
+  /** @type {string} */
+  #suiteName
+
+  /**
+   * @param {GlobalConfig} _globalConfig
+   * @param {Record<string, any>} reporterOptions
+   */
+  constructor(_globalConfig, reporterOptions) {
+    this.#suiteName = reporterOptions.suiteName ?? 'jest'
+  }
+
+  /**
+   * @type {Reporter["onRunStart"]}
+   */
+  onRunStart() {
+    this.tcServiceMessage('testSuiteStarted', { name: this.#suiteName })
+  }
+
+  /**
+   * @type {Reporter["onRunComplete"]}
+   */
+  onRunComplete() {
+    this.tcServiceMessage('testSuiteFinished', { name: this.#suiteName })
+  }
+
+  /**
+   * @type {Reporter["getLastError"]}
+   */
+  getLastError() {}
+
+  /**
+   * @type {Reporter["onTestFileResult"]}
+   * @param {Test} test
+   * @param {TestResult} testFileResult
+   */
+  onTestFileResult(test, testFileResult) {
+    // A flag to assert the test file result was actually reported.
+    let testFileResultWasReported = false
+
+    // Get the test file path relative to the cwd.
+    const testFilePath = path
+      .relative(test.context.config.cwd, testFileResult.testFilePath)
+      .replace(/\\/g, '/')
+
+    // Convert the test file path into a Teamcity-compatible test name prefix.
+    const tcTestPrefix = this.createTcTestPrefix(testFilePath)
+
+    // Gather console messages from the test file.
+    const consoleMessages = testFileResult.console?.map((logEntry) => {
+      return `console.${logEntry.type}\n  ${logEntry.message}\n${logEntry.origin}`
```

Comment:
Should `logEntry.origin` be indented too?

Reply:
nope, it's already indented. ex.:
```
console.error
  locale string requires a key!
    at getLocale (/home/ubuntu/workspace/teamcity/work/issues/30125/jest-tc-reporter/src/brave/components/common/locale.ts:15:13)
    at getTransactionStatusString (/home/ubuntu/workspace/teamcity/work/issues/30125/jest-tc-reporter/src/brave/components/brave_wallet_ui/utils/tx-utils.ts:189:19)
...
```